### PR TITLE
Add Strings.InStr and Strings.InStrRev methods (#31181)

### PR DESCRIPTION
**Pull request by EdwinEngelen**
I ported `Strings.InStr` and `Strings.InStrRev`. Part of issue #31181.

### CODE REVIEW

**Comment by danmosemsft** [comment]
Our tests run against .NET Framework also, as a check. In this case some have failed. You can click on the Details link above, it will eventually take you to:
https://mc.dot.net/#/user/EdwinEngelen/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Fmaster~2F/test~2Ffunctional~2Fcli~2F/43c057f16088aef0e501f89474ad9e19c4c7d9c8/workItem/Microsoft.VisualBasic.Tests

```
Assert.Equal() Failure
Expected: Start
Actual:   (null)
Stack Trace :
   at System.AssertExtensions.Throws[T](String paramName, Func`1 testCode) in D:\j\workspace\windows-TGrou---2a8f9c29\src\CoreFx.Private.TestUtilities\src\System\AssertExtensions.cs:line 88
```

This indicates that the ARgumentException parameter field is "Start" on .NET Core and null on .NET Framework. .NET Core is an improvement. To avoid this failure, change to a different AssertExtensions.Throws overload that has a parameter for the .NET Framework parameter name -- you can pass null to that -- and for .NET Core -- you can pass "Start" to that:

https://github.com/dotnet/corefx/blob/master/src/CoreFx.Private.TestUtilities/src/System/AssertExtensions.cs#L23

You can verify by running tests against .NET Framework locally yourself. This explains how -- look for `-target:netfx`

https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/developer-guide.md

**Comment by danmosemsft** [comment]
@EdwinEngelen we usually just push another commit when addressing PR feedback so it's not necessary to read all the code again. We squash on submit.

**Comment by danmosemsft** [comment]
Although when I catch up with master I personally use rebase rather than merge...at least that keeps separate changes.

**Comment by danmosemsft** [comment]
If you can resolve the conflicts we can merge  the PR.

File: src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/Strings.vb [hunk 1]
```diff
@@ -13,6 +13,11 @@ Namespace Global.Microsoft.VisualBasic
 
     Public Module Strings
 
+        Private Const STANDARD_COMPARE_FLAGS As CompareOptions =
+            CompareOptions.IgnoreCase Or CompareOptions.IgnoreWidth Or CompareOptions.IgnoreKanaType
+
+        Friend ReadOnly m_InvariantCompareInfo As CompareInfo = CultureInfo.InvariantCulture.CompareInfo
```
[the remark this answers is not in the document]
Comment (EdwinEngelen) [reply src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/Strings.vb]: It's just copy-paste of the 'old' code. We don't cleanup or refactor anything. That hasn't got any added value either. It's also easier when comparing NetFX and CoreFX source files.
What adds value is the tests. I'd rather invest time in creating good tests than in changing the code.

File: src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/Strings.vb [hunk 2]
```diff
@@ -229,6 +234,108 @@ Namespace Global.Microsoft.VisualBasic
...
 
+        Public Function InStr(ByVal String1 As String, ByVal String2 As String, <Microsoft.VisualBasic.CompilerServices.OptionCompareAttribute()> Optional ByVal [Compare] As CompareMethod = CompareMethod.Binary) As Integer
+            If Compare = CompareMethod.Binary Then
+                Return (InternalInStrBinary(0, String1, String2) + 1)
+            Else
+                Return (InternalInStrText(0, String1, String2) + 1)
+            End If
+        End Function
+
+        Public Function InStr(ByVal Start As Integer, ByVal String1 As String, ByVal String2 As String, <Microsoft.VisualBasic.CompilerServices.OptionCompareAttribute()> Optional ByVal [Compare] As CompareMethod = CompareMethod.Binary) As Integer
```
[the remark this answers is not in the document]
Comment (EdwinEngelen) [reply src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/Strings.vb]: See previous comment. It's a straightforward copy-paste.

File: src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/Strings.vb [hunk 3]
```diff
@@ -229,6 +234,108 @@ Namespace Global.Microsoft.VisualBasic
...
+            If Compare = CompareMethod.Binary Then
+                Return (InternalInStrBinary(Start - 1, String1, String2) + 1)
+            Else
+                Return (InternalInStrText(Start - 1, String1, String2) + 1)
+            End If
+        End Function
+
+        'THIS FUNCTION IS ZERO BASED
+        Private Function InternalInStrBinary(ByVal StartPos As Integer, ByVal sSrc As String, ByVal sFind As String) As Integer
+            Dim SrcLength As Integer
```
[the remark this answers is not in the document]
Comment (EdwinEngelen) [reply src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/Strings.vb]: And that breaks compatibility. People might use named arguments, which would no longer work (in C#) when the casing changes.

File: src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/Strings.vb [hunk 4]
```diff
@@ -229,6 +234,108 @@ Namespace Global.Microsoft.VisualBasic
...
+
+            If sSrc IsNot Nothing Then
+                SrcLength = sSrc.Length
+            Else
+                SrcLength = 0
+            End If
+
+            If StartPos > SrcLength OrElse SrcLength = 0 Then
```
[the remark this answers is not in the document]
Comment (EdwinEngelen) [reply src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/Strings.vb]: See other comment.

File: src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/Strings.vb [hunk 5]
```diff
@@ -229,6 +234,108 @@ Namespace Global.Microsoft.VisualBasic
...
+            End If
+
+            If (sFind Is Nothing) OrElse (sFind.Length = 0) Then
+                Return lStartPos
+            End If
+
+            Return GetCultureInfo().CompareInfo.IndexOf(sSrc, sFind, lStartPos, STANDARD_COMPARE_FLAGS)
+        End Function
+
+        Public Function InStrRev(ByVal StringCheck As String, ByVal StringMatch As String, Optional ByVal Start As Integer = -1, <Microsoft.VisualBasic.CompilerServices.OptionCompareAttribute()> Optional ByVal [Compare] As CompareMethod = CompareMethod.Binary) As Integer
```
[the remark this answers is not in the document]
Comment (EdwinEngelen) [reply src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/Strings.vb]: See other comments.

File: src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/Strings.vb (hunk 1 above)
[the remark this answers is not in the document]
Comment (Anipik) [reply src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/Strings.vb]: cc @danmosemsft do we have to skip the guidelines or refactoring here ?

File: src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/Strings.vb (hunk 3 above)
[the remark this answers is not in the document]
Comment (Anipik) [reply src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/Strings.vb]: This is private function. It wont be called be outside

**Comment by danmosemsft** [comment]
@anipik yeah we are leaving the sources pretty much verbatim except maybe comments, license, obviusly dead code. Also yes changing case of parameter names is breaking and VB usually has leading upper case. For the ports is mostly about having reasonable tests.

File: src/Microsoft.VisualBasic/tests/StringsTests.cs [hunk 6]
```diff
@@ -174,12 +174,175 @@ public void Filter_Objects_WhenObjectCannotBeConvertedToString_ThrowsArgumentOut
...
+        [InlineData("aa", null, 2)]
+        [InlineData("aa", "a", 2)]
+        [InlineData("aab", "a", 2)]
+        [InlineData("aba", "a", 3)]
+        [InlineData("ab", "a", 0)]
+        [InlineData("aa", "ab", 0)]
+        [InlineData("abab", "ab", 3)]
+        public void InStr_FromWithin(string string1, string string2, int expected)
+        {
+            var startPos = 2;
```
Comment (Anipik) [inline src/Microsoft.VisualBasic/tests/StringsTests.cs]: can you please remove var from all of these.
https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md
```We only use var when it's obvious what the variable type is (e.g. var stream = new FileStream(...) not var stream = OpenStandardInput()).```

File: src/Microsoft.VisualBasic/tests/StringsTests.cs [hunk 7]
```diff
@@ -174,12 +174,175 @@ public void Filter_Objects_WhenObjectCannotBeConvertedToString_ThrowsArgumentOut
...
             Assert.Equal(includeExpected, Strings.Filter(source, match, Include: true));
             Assert.Equal(excludeExpected, Strings.Filter(source, match, Include: false));
         }
 
+        [Theory]
+        [InlineData(null, null, 0)]
+        [InlineData(null, "", 0)]
+        [InlineData("", null, 0)]
+        [InlineData("", "", 0)]
+        [InlineData(null, "a", 0)]
```
Comment (Anipik) [inline src/Microsoft.VisualBasic/tests/StringsTests.cs]: These 5 values are common in some tests. we can move them to a function and supply them to the test using MemberData(nameof(function))
https://github.com/dotnet/corefx/blob/master/src/Common/tests/Tests/System/IO/PathInternal.Tests.cs#L193

Comment (Anipik) [reply src/Microsoft.VisualBasic/tests/StringsTests.cs]: Its generally nice to use memberData attribute when a lot of inline Data elements are present. I will recommend it for ```InStr_FromBegin```, ```InStr_FromWithin```, ```InStrRev_FromEnd``` and ```InStrRev_FromWithin```

**Comment by danmosemsft** [comment]
Remaining feedback on the tests above.

**Comment by karelz** [comment]
Is there more feedback response expected from @EdwinEngelen?

**Comment by Anipik** [comment]
There is some test code refactoring left here.

**Comment by EdwinEngelen** [comment]
I'm on holiday now. Hope to finish this next week.

**Comment by karelz** [comment]
@EdwinEngelen any update?

**Comment by EdwinEngelen** [comment]
I think I'm ready. 😄

**Comment by EdwinEngelen** [comment]
Ready for final code review @karelz

File: src/Microsoft.VisualBasic/tests/StringsTests.cs [hunk 8]
```diff
@@ -165,21 +165,141 @@ public void Filter_MultipleElements(string[] source, string match, string[] incl
...
+            Assert.Equal(expected, Strings.InStr(string1, string2));
+            Assert.Equal(expected, Strings.InStr(1, string1, string2));
+        }
+
+        [Theory]
+        [MemberData(nameof(InStr_TestData_NullsAndEmpties))]
+        [MemberData(nameof(InStr_FromWithin_TestData))]
+        public void InStr_FromWithin(string string1, string string2, int expected)
+        {
+            int startPos = 2;
```
Comment (Anipik) [inline src/Microsoft.VisualBasic/tests/StringsTests.cs]: we can inline the this variable. no need to define it separately
``` Assert.Equal(expected, Strings.InStr(2, string1, string2));```

File: src/Microsoft.VisualBasic/tests/StringsTests.cs [hunk 9]
```diff
@@ -165,21 +165,141 @@ public void Filter_MultipleElements(string[] source, string match, string[] incl
...
+        }
+
+        [Theory]
+        [InlineData(2)]
+        [InlineData(3)]
+        public void InStr_WhenStartGreatherThanLength_ReturnsZero(int start)
+        {
+            string string1 = "a";
+            string string2 = "a";
+            int expected = 0;
```
Comment (Anipik) [inline src/Microsoft.VisualBasic/tests/StringsTests.cs]: no need to define these variable, you can directly provide them in function call

File: src/Microsoft.VisualBasic/tests/StringsTests.cs [hunk 10]
```diff
@@ -165,21 +165,141 @@ public void Filter_MultipleElements(string[] source, string match, string[] incl
...
+            Assert.Equal(expected, Strings.InStr(start, string1, string2));
+        }
+
+        [Theory]
+        [InlineData(0)]
+        [InlineData(-1)]
+        public void InStr_WhenStartZeroOrLess_ThrowsArgumentException(int start)
+        {
+            string string1 = "a";
+            string string2 = "a";
```
Comment (Anipik) [inline src/Microsoft.VisualBasic/tests/StringsTests.cs]: same for these variables

File: src/Microsoft.VisualBasic/tests/StringsTests.cs [hunk 11]
```diff
@@ -165,21 +165,141 @@ public void Filter_MultipleElements(string[] source, string match, string[] incl
...
+            Assert.Equal(expected, Strings.InStrRev(stringCheck, stringMatch, start, CompareMethod.Text));
+        }
+
+        [Fact]
+        public void InStrRev_WhenStartMinusOne_SearchesFromEnd()
+        {
+            string stringCheck = "aa";
+            string stringMatch = "a";
+            int start = -1;
+            int expected = 2;
```
Comment (Anipik) [inline src/Microsoft.VisualBasic/tests/StringsTests.cs]: same here

File: src/Microsoft.VisualBasic/tests/StringsTests.cs [hunk 12]
```diff
@@ -165,21 +165,141 @@ public void Filter_MultipleElements(string[] source, string match, string[] incl
...
+        }
+
+        [Theory]
+        [InlineData(0)]
+        [InlineData(-2)]
+        [InlineData(-3)]
+        public void InStrRev_WhenStartZeroOrMinusTwoOrLess_ThrowsArgumentException(int start)
+        {
+            string stringCheck = "a";
+            string stringMatch = "a";
```
Comment (Anipik) [inline src/Microsoft.VisualBasic/tests/StringsTests.cs]: same here

File: src/Microsoft.VisualBasic/tests/StringsTests.cs [hunk 13]
```diff
@@ -165,21 +165,141 @@ public void Filter_MultipleElements(string[] source, string match, string[] incl
...
+        }
+
+        [Theory]
+        [InlineData(2)]
+        [InlineData(3)]
+        public void InStrRev_WhenStartGreatherThanLength_ReturnsZero(int start)
+        {
+            string stringCheck = "a";
+            string stringMatch = "a";
+            int expected = 0;
```
Comment (Anipik) [inline src/Microsoft.VisualBasic/tests/StringsTests.cs]: same here

**Comment by karelz** [comment]
@EdwinEngelen do you plan to respond to the code review feedback from @Anipik?

**Comment by EdwinEngelen** [comment]
I've inlined the variables. Comments of @Anipik resolved.

File: src/Microsoft.VisualBasic/tests/StringsTests.cs [hunk 14]
```diff
@@ -301,5 +406,67 @@ public void Trim_Valid(string str, string expected)
...
+            { null, "a", 0 },
+            { "a", null, 1 },
+            { "a", "a", 1 },
+            { "aa", "a", 1 },
+            { "ab", "a", 1 },
+            { "ba", "a", 2 },
+            { "b", "a", 0 },
+            { "a", "ab", 0 },
+            { "ab", "ab", 1 },
+        };
```
Comment (cston) [inline src/Microsoft.VisualBasic/tests/StringsTests.cs]: >} [](start = 8, length = 1)

Consider testing empty and non-empty strings for `InStr` and `InStrRev`:
```
{ "", "a", 0 },
{ "a", "", 0 },
```

**Comment by cston** [comment]
Thanks @EdwinEngelen! The PR has been merged.

**Pull request merged**